README.md
11/05/2021

<!-- README.md is generated from README.Rmd. Please edit that file -->
rsimpop
=======

<!-- badges: start -->
<!-- badges: end -->
Installation
------------

You can install rsimpop like so:

``` r
devtools::install_github("NickWilliamsSanger/rsimpop")
```

rsimpop
-------

This package facilitates the simultaneous simulation of multiple cellular compartments each with their own target population size and potentially also sub-compartments with differential fitness (driver compartments).

Simulate from Zygote for 1 year and subsample tree
--------------------------------------------------

``` r

##Initialise with seed (R and rsimpop separately)
SEED=37774323
initSimPop(SEED,bForce = TRUE)
#> NULL
##Setup a single compartment with a target pop of 50K
cfg=getDefaultConfig(target_pop_size  = 5e4,ndriver = 1,basefit = 0.2,rate = 0.1)
print(cfg)
#> $compartment
#>   val rate popsize      desc
#> 1   0 -1.0       1  outgroup
#> 2   1  0.1   50000 cellType1
#> 
#> $info
#>   val population fitness id driver1
#> 1   0          0       0  0       0
#> 2   1          0       0  0       0
#> 
#> $drivers
#>   val driver fitness
#> 1   1      1       0
##Simulate for 2years..
sp=sim_pop(NULL,params=list(n_sim_days=365*2,b_stop_at_pop_size=1),cfg=cfg)
#> n_sim_days: 730
#> b_stop_if_empty: 0
#> b_stop_at_pop_size: 1
#> maxt: 0
#> driver_rate_per_cell_per_day: 0
#> MAX_EVENTS= 1460 
#> MAX_SIZE= 150003
##Look at the population size trajectory
plot(sp)
```

<img src="man/figures/README-zygote-1.png" width="100%" />

``` r
##Subsample tree
sampledtree1=get_subsampled_tree(sp,100)
#> Starting checking the validity of tmp...
#> Found number of tips: n = 101 
#> Found number of nodes: m = 100 
#> Done.
print(sampledtree1)
#> 
#> Phylogenetic tree with 101 tips and 100 internal nodes.
#> 
#> Tip labels:
#>  s1, s2, s3, s4, s5, s6, ...
#> 
#> Rooted; includes branch lengths.
plot_tree(sampledtree1,cex.label = 0.5)
#> 
#> Phylogenetic tree with 101 tips and 100 internal nodes.
#> 
#> Tip labels:
#>  s1, s2, s3, s4, s5, s6, ...
#> 
#> Rooted; includes branch lengths.
title("Sampled Zygote Tree: Division Tree")
```

<img src="man/figures/README-zygote-2.png" width="100%" />

Notice how the sampled tree has 101 tips rather than the specified 100. This is because the simulator now always maintains an inactive outgroup (here s1). A group is rendered inactive by specifying a negative "rate" in the cfg$compartment dataframe. The tree branch lengths are now given in terms of the number of self renewal divisions. This allows the user to flexibly apply their own mutation acquisition model:

``` r
get_elapsed_time_tree
#> function (tree, mutrateperdivision = NULL, backgroundrate = NULL, 
#>     odf = 1) 
#> {
#>     N = length(tree$tip.label) + 1
#>     L = length(tree$edge.length)
#>     TT = max(tree$timestamp)
#>     idx.child = match(tree$edge[, 2], tree$edge[, 1])
#>     duration = ifelse(is.na(idx.child), TT - tree$tBirth, tree$tBirth[idx.child] - 
#>         tree$tBirth)
#>     duration[which(tree$state == 0)] = 0
#>     if (!is.null(mutrateperdivision)) {
#>         if (odf > 1) {
#>             tree$edge.length = sapply(tree$ndiv, function(n) sum(rpois(n, 
#>                 mutrateperdivision))) + get_nb(n = L, meanmuts = backgroundrate * 
#>                 duration, od = odf)
#>         }
#>         else {
#>             tree$edge.length = sapply(tree$ndiv, function(n) sum(rpois(n, 
#>                 mutrateperdivision))) + rpois(L, backgroundrate * 
#>                 duration)
#>         }
#>     }
#>     else {
#>         tree$edge.length = duration
#>     }
#>     tree
#> }
#> <bytecode: 0x7fc324f0c288>
#> <environment: namespace:rsimpop>
sampledtree1m=get_elapsed_time_tree(sampledtree1,mutrateperdivision=1,backgroundrate=15/365)
plot_tree(sampledtree1m,cex.label = 0.5);title("Sampled Zygote Tree: Mutation Tree")
#> 
#> Phylogenetic tree with 101 tips and 100 internal nodes.
#> 
#> Tip labels:
#>  s1, s2, s3, s4, s5, s6, ...
#> 
#> Rooted; includes branch lengths.
```

<img src="man/figures/README-mutacq-1.png" width="100%" />

Actually this illustrates a potential problem with the outgroup sample still having a acquired mutations because it has a finite duration (0 to 365 days)..

The changes between compartments is specified in a separate data.frame, tree$events , that is maintained and updated by the simulator.

``` r
t1=plot_tree(sampledtree1m,cex.label = 0.5);title("Sampled Zygote Tree: Mutation Tree")
node_labels(t1,cex=0.5)
```

<img src="man/figures/README-events-1.png" width="100%" />

``` r
print(sampledtree1m$events)
#>   value driverid node ts
#> 1     0        0    1  0
#> 2     1        0  103  0
```

Notice how the events dataframe specifies the compartment for the outgroup and the rest of the tree.

We can introduce another cell compartment as follows:

``` r
cfg=sampledtree1$cfg
cfg=addCellCompartment(cfg,population = 5e4,rate=1/50,ndriver=1,descr="MyTissue",basefit = 0.3)
cfg$compartment$rate[2]=1/120  ## change the rate of compartment 1
sampledtree1a=addDifferentiationEvents(sampledtree1,cfg,2,nEvent=10)
print(sampledtree1a$events)
#>    value driverid node       ts
#> 1      0        0    1   0.0000
#> 2      1        0  103   0.0000
#> 3      2        0    4 135.4454
#> 4      2        0    8 135.4454
#> 5      2        0   11 135.4454
#> 6      2        0   14 135.4454
#> 7      2        0   38 135.4454
#> 8      2        0   54 135.4454
#> 9      2        0   80 135.4454
#> 10     2        0   81 135.4454
#> 11     2        0   90 135.4454
#> 12     2        0   96 135.4454
```

Each branch carries its final compartment membership in the "state" vector.

``` r
sampledtree1a$color = c("grey","black","red")[sampledtree1a$state+1]
plot_tree(sampledtree1a,cex.label = 0.5);title("Highlights branches with compartment changes")
#> 
#> Phylogenetic tree with 101 tips and 100 internal nodes.
#> 
#> Tip labels:
#>  s1, s2, s3, s4, s5, s6, ...
#> 
#> Rooted; includes branch lengths.
```

<img src="man/figures/README-plotcompartment-1.png" width="100%" /> The above plot does not capture the situation when compartment changes take place mid-branch - so alternatively we can better visualise the situation using the built in function *plot\_tree\_events*

``` r
plot_tree_events(sampledtree1a)
```

<img src="man/figures/README-plotc2-1.png" width="100%" />

    #>   uval     col pch
    #> 1  0:0 #E41A1C  19
    #> 2  1:0 #377EB8  19
    #> 3  2:0 #4DAF4A  19
    #>   uval     col pch
    #> 1  0:0 #E41A1C  19
    #> 2  1:0 #377EB8  19
    #> 3  2:0 #4DAF4A  19
    #>    value driverid node       ts fitness key     col pch idx
    #> 1      0        0    1   0.0000       0 0:0 #E41A1C  19   1
    #> 2      1        0  103   0.0000       0 1:0 #377EB8  19   2
    #> 3      2        0    4 135.4454       0 2:0 #4DAF4A  19  11
    #> 4      2        0    8 135.4454       0 2:0 #4DAF4A  19  16
    #> 5      2        0   11 135.4454       0 2:0 #4DAF4A  19  24
    #> 6      2        0   14 135.4454       0 2:0 #4DAF4A  19  31
    #> 7      2        0   38 135.4454       0 2:0 #4DAF4A  19  79
    #> 8      2        0   54 135.4454       0 2:0 #4DAF4A  19 110
    #> 9      2        0   80 135.4454       0 2:0 #4DAF4A  19 161
    #> 10     2        0   81 135.4454       0 2:0 #4DAF4A  19 165
    #> 11     2        0   90 135.4454       0 2:0 #4DAF4A  19 182
    #> 12     2        0   96 135.4454       0 2:0 #4DAF4A  19 193
    #> 
    #> Phylogenetic tree with 101 tips and 100 internal nodes.
    #> 
    #> Tip labels:
    #>  s1, s2, s3, s4, s5, s6, ...
    #> 
    #> Rooted; includes branch lengths.

We've already updated the config with the target population sizes and division rates, so we're ready to simulate:

``` r
sp2=sim_pop(sampledtree1a,params=list(n_sim_days=365*10),cfg=sampledtree1a$cfg)
#> n_sim_days: 3650
#> b_stop_if_empty: 0
#> b_stop_at_pop_size: 0
#> maxt: 135.445371986415
#> driver_rate_per_cell_per_day: 0
#> MAX_EVENTS= 7300 
#> MAX_SIZE= 300003
sp2=combine_simpops(sp,sp2)
plot(sp2)
```

<img src="man/figures/README-simagain-1.png" width="100%" />

``` r
sampledtree2=get_subsampled_tree(sp2,100)
#> Starting checking the validity of tmp...
#> Found number of tips: n = 101 
#> Found number of nodes: m = 100 
#> Done.
plot_tree_events(sampledtree2,cex.label = 0.5)
```

<img src="man/figures/README-simagain-2.png" width="100%" />

    #>   uval     col pch
    #> 1  0:0 #E41A1C  19
    #> 2  1:0 #377EB8  19
    #> 3  2:0 #4DAF4A  19
    #>   uval     col pch
    #> 1  0:0 #E41A1C  19
    #> 2  1:0 #377EB8  19
    #> 3  2:0 #4DAF4A  19
    #>   value driverid node       ts fitness key     col pch idx
    #> 1     0        0    1   0.0000       0 0:0 #E41A1C  19   1
    #> 2     1        0  103   0.0000       0 1:0 #377EB8  19   2
    #> 3     2        0  107 135.4454       0 2:0 #4DAF4A  19   6
    #> 4     2        0  122 135.4454       0 2:0 #4DAF4A  19  37
    #> 5     2        0  137 135.4454       0 2:0 #4DAF4A  19  68
    #> 6     2        0  144 135.4454       0 2:0 #4DAF4A  19  78
    #> 7     2        0  170 135.4454       0 2:0 #4DAF4A  19 133
    #> 8     2        0  188 135.4454       0 2:0 #4DAF4A  19 169
    #> 9     2        0  196 135.4454       0 2:0 #4DAF4A  19 186
    #> 
    #> Phylogenetic tree with 101 tips and 100 internal nodes.
    #> 
    #> Tip labels:
    #>  s1, s2, s3, s4, s5, s6, ...
    #> 
    #> Rooted; includes branch lengths.

Selection based simulation
--------------------------

Here we are interested in the simple situation of one cellular compartment with multiple sub-compartments.

``` r
run_selection_sim
#> function (initial_division_rate = 0.1, final_division_rate = 1/365, 
#>     target_pop_size = 1e+05, nyears_driver_acquisition = 15, 
#>     nyears = 40, fitness = 0.2, minprop = 0.001, mindriver = 1, 
#>     maxtry = 40) 
#> {
#>     cfg = getDefaultConfig(target_pop_size, rate = initial_division_rate, 
#>         ndriver = 1, basefit = fitness)
#>     params = list(n_sim_days = nyears_driver_acquisition * 365, 
#>         b_stop_at_pop_size = 1, b_stop_if_empty = 0)
#>     growthphase = sim_pop(NULL, params = params, cfg)
#>     ndivkeep = 0
#>     mdivkeep = 0
#>     gdivkeep = 0
#>     tree0 = get_tree_from_simpop(growthphase)
#>     if (growthphase$status == 0) {
#>         tree1 = tree0
#>         params[["n_sim_days"]] = nyears * 365
#>         params[["b_stop_if_empty"]] = 1
#>         dc = 0
#>         tries = 0
#>         tree1_tmp = tree1
#>         while (dc < max(minprop * target_pop_size, mindriver)) {
#>             if (tries >= maxtry) {
#>                 return(NULL)
#>             }
#>             cat("No driver found: tries=", tries, "\n")
#>             tries = tries + 1
#>             tree1_tmp = addDriverEvent(tree1, tree1$cfg, 1, fitness = fitness)
#>             print(tree1_tmp$cfg$info)
#>             params[["b_stop_at_pop_size"]] = 1
#>             adult2 = sim_pop(tree1_tmp, params = params, tree1_tmp$cfg)
#>             adult2a = combine_simpops(growthphase, adult2)
#>             tree2 = get_tree_from_simpop(adult2a)
#>             params[["b_stop_at_pop_size"]] = 0
#>             cfg = tree2$cfg
#>             cfg$compartment$rate[2] = final_division_rate
#>             cfg$compartment$popsize[2] = target_pop_size
#>             adult2 = sim_pop(tree2, params = params, cfg)
#>             adult2 = combine_simpops(adult2a, adult2)
#>             dc = adult2$cfg$info$population[3]
#>         }
#>     }
#>     else {
#>         gdivkeep = mean(nodeHeights(tree0)[which(tree0$edge[, 
#>             2] <= length(tree0$tip.label)), 2])
#>         cfg$compartment$rate[2] = final_division_rate
#>         cfg$compartment$popsize[2] = target_pop_size
#>         years = nyears
#>         params[["n_sim_days"]] = nyears_driver_acquisition * 
#>             365
#>         params[["b_stop_at_pop_size"]] = 0
#>         adult1 = sim_pop(tree0, params = params, cfg)
#>         adult1 = combine_simpops(growthphase, adult1)
#>         tree1 = get_tree_from_simpop(adult1)
#>         params[["n_sim_days"]] = nyears * 365
#>         params[["b_stop_if_empty"]] = 1
#>         dc = 0
#>         tries = 0
#>         tree1_tmp = tree1
#>         while (dc < max(minprop * target_pop_size, mindriver)) {
#>             if (tries >= maxtry) {
#>                 return(NULL)
#>             }
#>             cat("No driver found: tries=", tries, "\n")
#>             tries = tries + 1
#>             tree1_tmp = addDriverEvent(tree1, tree1$cfg, 1, fitness = fitness)
#>             if (TRUE) {
#>                 ndivkeep = nodeHeights(tree1_tmp)[which(tree1_tmp$edge[, 

CHANGES.md
# CHANGES

## 2.1.2

* run_neutral_trajectory now works when initialised with a non-null simpop object.

## 2.1.1

* Added missing 2.1.0  changes to CHANGES.md

## 2.1.0

* Removal of redundant tree plotting functions

## 2.0.4

* Switched from multiplicative to additive selection model for compound drivers
* Added parameter to plot_tree_events to show selection coefficient on legend.

## 2.0.3

* Fixed bug in enhanced handling of drivers.

## 2.0.2

* Enhanced handling of drivers so that there is no set cap on the number of drivers.

## 2.0.1

* Initial Release


R/wrapped_sims.R
### This file contains wrapped calls to the simulator that serve as examples

#' Runs a single compartment/single driver scenario
#'
#' The driver is subject to stochastic extinction and so is repeatedly dropped in until it "takes".
#' The prevailing state when the driver is introduced is saved and reinstated with each attempted introduction.
#' @param initial_division_rate - Rate of symmetric cell division during development
#' @param final_division_rate - Rate of symmetric cell division once population equilibrium is reached.
#' @param target_pop_size - Size of target population
#' @param nyears_driver_acquisition - When driver is acquired.
#' @param nyears - Total number of years to run the simulation
#' @param fitness - relative fitness advantage.  Cells carrying this divide at a rate=(1+fitness)*baserate
#' @param minprop - Minimum aberrant cell fraction at nyears for driver to be regarded as have "taken"
#' @param maxtry  - Maximum number of attempts to introduce a driver
#' @return simpop object.
#' @export
#' @examples
#' selsim=run_selection_sim(0.05,1/365,target_pop_size = 5e4,nyears = 50,fitness=0.3)
run_selection_sim=function( initial_division_rate=0.1,
                            final_division_rate=1/365,
                            target_pop_size=1e5,
                            nyears_driver_acquisition=15,
                            nyears=40,
                            fitness=0.2, ## relative fitness
                            minprop=0.001,
                            mindriver=1,
                            maxtry=40
){
  cfg=getDefaultConfig(target_pop_size,rate=initial_division_rate,ndriver=1,basefit = fitness)
  params=list(n_sim_days=nyears_driver_acquisition*365,
              b_stop_at_pop_size=1,
              b_stop_if_empty=0
  )
  growthphase=sim_pop(NULL,params=params,cfg)
  ndivkeep=0
  mdivkeep=0
  gdivkeep=0
  tree0=get_tree_from_simpop(growthphase)
  if(growthphase$status==0){
    ##We've exited before maturity and need to add driver now..
    tree1=tree0
    params[["n_sim_days"]]=nyears*365 ##years of simulation
    params[["b_stop_if_empty"]]=1

    dc=0
    tries=0
    tree1_tmp=tree1
    while(dc<max(minprop*target_pop_size,mindriver) ){
      if(tries>=maxtry){
        return(NULL)
      }
      cat("No driver found: tries=",tries,"\n")
      tries=tries+1
      tree1_tmp=addDriverEvent(tree1,tree1$cfg,1,fitness=fitness)
      print(tree1_tmp$cfg$info)
      params[["b_stop_at_pop_size"]]=1
      adult2=sim_pop(tree1_tmp,params=params,tree1_tmp$cfg)
      adult2a=combine_simpops(growthphase,adult2)
      tree2=get_tree_from_simpop(adult2a)
      params[["b_stop_at_pop_size"]]=0
      cfg=tree2$cfg
      cfg$compartment$rate[2]=final_division_rate
      cfg$compartment$popsize[2]=target_pop_size
      adult2=sim_pop(tree2,params=params,cfg)
      adult2=combine_simpops(adult2a,adult2)
      dc=adult2$cfg$info$population[3]
    }
  }else{
    gdivkeep=mean(nodeHeights(tree0)[which(tree0$edge[,2]<=length(tree0$tip.label)),2])
    cfg$compartment$rate[2]=final_division_rate
    cfg$compartment$popsize[2]=target_pop_size
    years=nyears
    params[["n_sim_days"]]=nyears_driver_acquisition*365 ##years of simulation
    params[["b_stop_at_pop_size"]]=0 ## So it doesn't stop simulating immediately
    adult1=sim_pop(tree0,params=params,cfg)
    adult1=combine_simpops(growthphase,adult1)
    tree1=get_tree_from_simpop(adult1)
    params[["n_sim_days"]]=nyears*365 ##years of simulation
    params[["b_stop_if_empty"]]=1
    dc=0
    tries=0
    tree1_tmp=tree1
    while(dc<max(minprop*target_pop_size,mindriver) ){
      if(tries>=maxtry){
        return(NULL)
      }
      cat("No driver found: tries=",tries,"\n")
      tries=tries+1
      tree1_tmp=addDriverEvent(tree1,tree1$cfg,1,fitness=fitness)
      if(TRUE){
        ndivkeep=nodeHeights(tree1_tmp)[which(tree1_tmp$edge[,2]==tree1_tmp$events$node[3]),2]
        mdivkeep=mean(nodeHeights(tree1_tmp)[which(tree1_tmp$edge[,2]<=length(tree1_tmp$tip.label)),2])
      }
      print(tree1_tmp$cfg$info)
      adult2=sim_pop(tree1_tmp,params=params,tree1_tmp$cfg)
      dc=adult2$cfg$info$population[3]
    }
    adult2=combine_simpops(adult1,adult2)
  }

  fulltree=get_tree_from_simpop(adult2)
  fulltree$tries=tries
  fulltree$ndivkeep=ndivkeep
  fulltree$mdivkeep=mdivkeep
  fulltree$gdivkeep=gdivkeep
  return(fulltree)
}

#' Runs a multiple driver scenario
#'
#' The user specifies the rate at which drivers are introduced and the distribution the selective coefficients are drawn from.
#' Note that unlike in \code{\link{run_selection_sim}} the drivers are allowed to stochastically die out.
#' The drivers at the specified rate with the gap between successive introductions exponentially distributed.
#' @param initial_division_rate - Rate of symmetric cell division during development
#' @param final_division_rate - Rate of symmetric cell division once population equilibrium is reached.
#' @param target_pop_size - Size of target population
#' @param drivers_per_year - The expected number of drivers per year
#' @param nyears - Total number of years to run the simulation
#' @param fitnessGen - Function that returns a single value for the selective coefficient
#' @param bForceReseed - Whether to reseed.
#' @param offset - time offset for choosing seed.
#' @return simpop object.
#' @export
#' @examples
#' fitnessGen=function(){
#'  trials=rexp(100,rate=30)
#'  idx=which(trials>0.05)
#'  if(length(idx)==0){
#'   fitnessGen()
#'  }else{
#'   trials[idx[1]]
#'  }
#' }
#' dps=run_driver_process_sim(0.1,1/(2*190),target_pop_size = 1e5,nyears = 30,fitness=fitnessGen,drivers_per_year = 5)
#'
run_driver_process_sim=function( initial_division_rate=0.1,
                                  final_division_rate=1/365,
                                  target_pop_size=1e5,
                                  drivers_per_year=0.1,
                                  nyears=40,
                                  fitnessGen=function(){0},
                                  bForceReseed=FALSE,offset=0
){
  if(bForceReseed){
    delay=(offset/10 +1)
    Sys.sleep(delay)
    cat("delay=",delay,"\n")
    initSimPop(-1,bForce = TRUE)
  }
  ##driverdt=cumsum(rexp(1000,drivers_per_year/365))
  dpcpd=(drivers_per_year/365)/target_pop_size

  k=0
  nsd=nyears*365
  cfg=getDefaultConfig(1e5,rate=initial_division_rate,ndriver=1,basefit = 0)
  params=list(n_sim_days=nsd,##This is the longest that the simulation will continue
              b_stop_at_pop_size=1,
              b_stop_if_empty=0,
              driver_rate_per_cell_per_day=dpcpd
  )
  growthphase=get_tree_from_simpop(sim_pop(NULL,params=params,cfg))
  while(growthphase$status==2 || growthphase$status==0){
    k=k+1
    params=list(n_sim_days=nsd,##This is the longest that the simulation will continue
                b_stop_at_pop_size=1,
                b_stop_if_empty=0,
                driver_rate_per_cell_per_day=dpcpd
    )
    if(max(growthphase$timestamp)>params$n_sim_days){
      stop("ts>n_sim_days:unexpected bahaviour")
    }
    growthphase=addDriverEvent(growthphase,growthphase$cfg,currentCompartment = 1,fitness=fitnessGen())
    gpk=get_tree_from_simpop(sim_pop(growthphase,params=params,growthphase$cfg))
    growthphase=combine_simpops(growthphase,gpk)
  }
  tree0=get_tree_from_simpop(growthphase)
  if(max(tree0$timestamp)>=nyears*365){
    tree0$n_introduced_drivers=k
    tree0
    return(tree0)
  }
  tree0$cfg$compartment$rate[2]=final_division_rate
  tree0$cfg$compartment$popsize[2]=target_pop_size
  years=nyears
  params[["b_stop_at_pop_size"]]=0 ## So it doesn't stop simulating immediately
  if(max(tree0$timestamp)>=nyears*365){
    return(tree0)
  }
  adult=sim_pop(tree0,params=params,tree0$cfg)
  adult=combine_simpops(tree0,adult)
  tree0=get_tree_from_simpop(adult)
  if(length(which(tree0$cfg$info$fitness==0 & tree0$cfg$info$population>0))>2){
    browser()
  }
  while(TRUE){
    
    ## Deal with end of last sim and next driver being too close together. Very rarely kicks in.
    ##params[["n_sim_days"]]=min(nyears*365,max(driverdt[k],max(tree0$timestamp)+0.1)) ##years of simulation
    if(max(tree0$timestamp)>=nyears*365){
      tree0$n_introduced_drivers=k
      tree0
      return(tree0)
    }
    k=k+1
    #browser()
    tree0=addDriverEvent(tree0,tree0$cfg,1,fitness=fitnessGen())
    tmp2=tree0
    adult=sim_pop(tree0,params=params,tree0$cfg,b_verbose = FALSE)
    tree0=combine_simpops(tree0,adult)
    #tree0=get_tree_from_simpop(adult)
    #print(tree0$cfg$info %>% dplyr::filter(population>0))
    if(length(which(tree0$cfg$info$fitness==0 & tree0$cfg$info$population>0))>2){
      browser()
    }
    if(k>100000){
      stop("too many iterations!")
    }
  }
  

}

#' Runs a single compartment/single driver scenario with transient selection
#'
#' The driver is subject to stochastic extinction and so is repeatedly dropped in until it "takes".
#' The prevailing state when the driver is introduced is saved and reinstated with each attempted introduction.
#' @param initial_division_rate - Rate of symmetric cell division during development
#' @param final_division_rate - Rate of symmetric cell division once population equilibrium is reached.
#' @param target_pop_size - Size of target population
#' @param nyears_driver_acquisition - When driver is acquired.
#' @param nyears_transient_end - When selective advantage is set to 0.
#' @param nyears - Total number of years to run the simulation
#' @param fitness - relative fitness advantage.  Cells carrying this divide at a rate=(1+fitness)*baserate
#' @param minprop - Minimum aberrant cell fraction at nyears for driver to be regarded as have "taken"
#' @param maxtry  - Maximum number of attempts to introduce a driver
#' @return simpop object.
#' @export
#' @examples
#' tselsim=run_transient_selection(0.05,1/365,target_pop_size = 5e4,nyears_driver_acquisition=15,
#' nyears_transient_end=30,nyears=50,fitness=0.5)
run_transient_selection=function( initial_division_rate,
                                  final_division_rate,
                                  target_pop_size=1e5,
                                  nyears_driver_acquisition=15,
                                  nyears_transient_end=30,
                                  nyears=40,
                                  fitness=0.2, ## relative fitness
                                  minprop=0.05
){
  selsim=run_selection_sim(initial_division_rate,
                                    final_division_rate,
                                    target_pop_size,
                                    nyears_driver_acquisition,
                                    nyears_transient_end,
                                    fitness,minprop)
  ##switch off
  cfg=selsim$cfg
  cfg$info$fitness[3]=0
  params=selsim$params
  params[["n_sim_days"]]=nyears*365
  params[["maxt"]]=NULL
  final=sim_pop(selsim,params=params,cfg)
  final=combine_simpops(selsim,final)
  return(get_tree_from_simpop(final))
}

#' Runs a simple neutral simulation
#'
#' @param initial_division_rate - Rate of symmetric cell division during development
#' @param final_division_rate - Rate of symmetric cell division once population equilibrium is reached.
#' @param target_pop_size - Size of target population
#' @param nyears - Total number of years to run the simulation
#' @return simpop object.
#' @export
#' @examples
#' neutsim=run_neutral_sim(0.05,1/365,target_pop_size = 5e4,nyears=80)
run_neutral_sim=function( initial_division_rate,
                                            final_division_rate,
                                            target_pop_size=1e5,
                                            nyears=40
){
  cfg=getDefaultConfig(target_pop_size,rate=initial_division_rate,ndriver=1,basefit = 0)
  params=list(n_sim_days=nyears*365,
              b_stop_at_pop_size=1,
              b_stop_if_empty=0
  )
  growthphase=sim_pop(NULL,params=params,cfg)
  cfg$compartment$rate[2]=final_division_rate
  cfg$compartment$popsize[2]=target_pop_size
  params[["b_stop_at_pop_size"]]=0
  adult1=sim_pop(growthphase,params=params,cfg)
  return(combine_simpops(growthphase,adult1))
}


add_driver=function(simpop,params,acceptance_threshold=0.05){
  tree1=get_tree_from_simpop(simpop)
  params[["b_stop_if_empty"]]=1
  tree1=simpop
  dc=0
  tries=0
  tree1_tmp=tree1
  while(dc/target_pop_size<0.05){
    cat("No driver found: tries=",tries,"\n")
    idx=sample(length(tree1$tip.label)-1,1)+1
    celltype=rep(NA,length(tree1$tip.label))
    celltype[idx]=-1
    tree1_tmp=assign_celltype(tree1,celltype,tree1$cfg)
    simpop2=sim_pop(tree1_tmp,params=params,tree1_tmp$cfg)

    dc=simpop2$cfg$info$population[3]
    #print(adult2$cfg$info)
    tries=tries+1
    if(tries>max_tries){
      stop("Unable to add driver.. Too many attempts")
    }
  }
  simpop=combine_simpops(simpop,simpop2)
  simpop
}

#' Runs a simple single compartment neutral simulation with a specified trajectory
#'
#' @param simpop - Rate of symmetric cell division during development
#' @param initial_division_rate - Rate of symmetric cell division once population equilibrium is reached.
#' @param trajectory - data.frame - with fields ts(timestamp in days),target_pop_size,division_rate
#' @param nyears - Total number of years to run the simulation
#' @return simpop object.
#' @export
#' @examples
#' trajectory=data.frame(ts=365*(1:80),target_pop_size=5e4+100*(1:80),division_rate=1/(2*190))
#' trajectory$target_pop_size[5:10]=2*trajectory$target_pop_size[5:10]
#' trajectory$target_pop_size[11:15]=0.2*trajectory$target_pop_size[11:15]
#' sp=run_neutral_trajectory(NULL,0.5,trajectory)
#' plot(sp)
run_neutral_trajectory=function(simpop,initial_division_rate,trajectory){
  if(any(trajectory$division_rate>1)){
    stop("Supplied division rate too high; should be less than 1.0")
  }
  if(any(trajectory$target_pop_size>20e6)){
    stop("Supplied population too high; should be less than 20e6")
  }
  params=list(n_sim_days=trajectory$ts[1],##This is the longest that the simulation will continue
              b_stop_at_pop_size=1,
              b_stop_if_empty=0
  )
  if(is.null(simpop)){
    cfg=getDefaultConfig(target_pop_size = trajectory$target_pop_size[1],rate = initial_division_rate,basefit = 0,ndriver = 1)

    
    sp=sim_pop(NULL,params=params,cfg,b_verbose = FALSE)
    idx=1
  }else{
    maxt=max(simpop$timestamp)
    idx=which(trajectory$ts>maxt)
    if(length(idx)>0){
      idx=idx[1]
    }else{
      idx=1
    }

    sp=simpop
  }
  params[["b_stop_at_pop_size"]]=0
  for(i in idx:(length(trajectory$ts)-1)){
    cfg=list(compartment=data.frame(val=c(0,1),rate=c(-1,trajectory$division_rate[i]),popsize=c(1,trajectory$target_pop_size[i])),
           info=sp$cfg$info)
    cfg=getDefaultConfig(target_pop_size = trajectory$target_pop_size[i],rate = trajectory$division_rate[i],basefit = 0,ndriver = 1)
    params[["n_sim_days"]]=trajectory$ts[i+1]
    spx=sim_pop(get_tree_from_simpop(sp),params=params,cfg,b_verbose = FALSE)
    sp=combine_simpops(sp,spx)
  }
  sp
}


run_2compartment_model=function(){
   ##Initialise LT-HSC
   #

}

simple_continuous_selection=function(){

}

#' Restarts a multiple driver scenario
#'
#' The user specifies the rate at which drivers are introduced and the distribution the selective coefficients are drawn from.
#' Note that unlike in \code{\link{run_selection_sim}} the drivers are allowed to stochastically die out.
#' The drivers at the specified rate with the gap between successive introductions exponentially distributed.
#' @param insim - Result of previous sim
#' @param final_division_rate - Rate of symmetric cell division once population equilibrium is reached.
#' @param target_pop_size - Size of target population
#' @param drivers_per_year - The expected number of drivers per year
#' @param nyears - Total number of years to run the simulation
#' @param bForceReseed - Whether to reseed.
#' @param offset - time offset for choosing seed.
#' @return simpop object.
#' @export
#' @examples
#' fitnessGen=function(){
#'  trials=rexp(100,rate=30)
#'  idx=which(trials>0.05)
#'  if(length(idx)==0){
#'   fitnessGen()
#'  }else{
#'   trials[idx[1]]
#'  }
#' }
#' dps=run_driver_process_sim(0.1,1/(2*190),target_pop_size = 1e5,nyears = 30,fitness=fitnessGen,drivers_per_year = 5)
#' ## Do stuff - subsample, plot trees etc and the restart to simulate the next 10 years:
#' dps2=continue_driver_process_sim(dps,40,fitnessGen = fitnessGen)
#' ## Do stuff and simulation the next 10 years
#' dps2=continue_driver_process_sim(dps2,50,fitnessGen = fitnessGen)
continue_driver_process_sim=function(insim,nyears,fitnessGen){
  tree0=insim
  params=insim$params
  params$maxt=NULL
  params$n_sim_days=nyears*365
  if(length(which(tree0$cfg$info$fitness==0 & tree0$cfg$info$population>0))>2){
    browser()
  }
  k=0
  while(TRUE){
    k=k+1
    ## Deal with end of last sim and next driver being too close together. Very rarely kicks in.
    ##params[["n_sim_days"]]=min(nyears*365,max(driverdt[k],max(tree0$timestamp)+0.1)) ##years of simulation
    if(max(tree0$timestamp)>=nyears*365){
      return(tree0)
    }
    tree0=addDriverEvent(tree0,tree0$cfg,1,fitness=fitnessGen())
    tmp2=tree0
    adult=sim_pop(tree0,params=params,tree0$cfg,b_verbose = FALSE)
    tree0=combine_simpops(tree0,adult)
    #tree0=get_tree_from_simpop(adult)
    #print(tree0$cfg$info %>% dplyr::filter(population>0))
    if(length(which(tree0$cfg$info$fitness==0 & tree0$cfg$info$population>0))>2){
      browser()
    }
    if(k>100000){
      stop("too many iterations!")
    }
  }
  ND=ifelse(is.null(tree0$n_introduced_drivers),0,tree0$n_introduced_drivers)
  tree0$n_introduced_drivers=ND+k
}



#' Converts annual growth selection coefficient to per division selection coefficient
#' @param S input annual selection coefficient
#' @param divrate division rate parameter specified in divisions per day.
#' @return s : per division selection coefficient to be supplied to rsimpop
#' @export
#'
convert_annualS_to_s=function(S,divrate){
  log(1 + S)/(365 * divrate)
}
